frame/fast-unstake/src/lib.rs

```diff
@@ -0,0 +1,500 @@
+				}
+			}
+
+			if <T as pallet_staking::Config>::ElectionProvider::ongoing() {
+				// NOTE: we assume `ongoing` does not consume any weight.
+				// there is an ongoing election -- we better not do anything. Imagine someone is not
+				// exposed anywhere in the last era, and the snapshot for the election is already
+				// taken. In this time period, we don't want to accidentally unstake them.
+				return T::DbWeight::get().reads(2)
+			}
+
+			let UnstakeRequest { stash, mut checked, maybe_pool_id } = match Head::<T>::take()
+				.or_else(|| {
+					// NOTE: there is no order guarantees in `Queue`.
+					Queue::<T>::drain()
+						.map(|(stash, maybe_pool_id)| UnstakeRequest {
+							stash,
+							maybe_pool_id,
+							checked: Default::default(),
+						})
+						.next()
+				}) {
+				None => {
+					// There's no `Head` and nothing in the `Queue`, nothing to do here.
+					return T::DbWeight::get().reads(4)
+				},
+				Some(head) => head,
+			};
+
+			log!(
+				debug,
+				"checking {:?}, eras_to_check_per_block = {:?}, remaining_weight = {:?}",
+				stash,
+				eras_to_check_per_block,
+				remaining_weight
+			);
+
+			// the range that we're allowed to check in this round.
+			let current_era = pallet_staking::CurrentEra::<T>::get().unwrap_or_default();
+			let bonding_duration = <T as pallet_staking::Config>::BondingDuration::get();
+			// prune all the old eras that we don't care about. This will help us keep the bound
+			// of `checked`.
+			checked.retain(|e| *e >= current_era.saturating_sub(bonding_duration));
+			let unchecked_eras_to_check = {
+				// get the last available `bonding_duration` eras up to current era in reverse
+				// order.
+				let total_check_range = (current_era.saturating_sub(bonding_duration)..=
+					current_era)
+					.rev()
+					.collect::<Vec<_>>();
+				debug_assert!(
+					total_check_range.len() <= (bonding_duration + 1) as usize,
+					"{:?}",
+					total_check_range
+				);
+
+				// remove eras that have already been checked, take a maximum of
+				// eras_to_check_per_block.
+				total_check_range
+					.into_iter()
+					.filter(|e| !checked.contains(e))
+					.take(eras_to_check_per_block as usize)
+					.collect::<Vec<_>>()
+			};
+
+			log!(
+				debug,
+				"{} eras to check: {:?}",
+				unchecked_eras_to_check.len(),
+				unchecked_eras_to_check
+			);
+
+			if unchecked_eras_to_check.is_empty() {
+				// `stash` is not exposed in any era now -- we can let go of them now.
+				let num_slashing_spans = Staking::<T>::slashing_spans(&stash).iter().count() as u32;
+
+				let ctrl = match pallet_staking::Bonded::<T>::get(&stash) {
+					Some(ctrl) => ctrl,
+					None => {
+						Self::deposit_event(Event::<T>::Errored { stash });
+						return <T as Config>::WeightInfo::on_idle_unstake()
+					},
+				};
+
+				let ledger = match pallet_staking::Ledger::<T>::get(ctrl) {
+					Some(ledger) => ledger,
+					None => {
+						Self::deposit_event(Event::<T>::Errored { stash });
```

Comment:
This isn't expected to happen right? Should we log these?

Reply:
> This isn't expected to happen right? Should we log these?
Since there is an event deposited for that - probably not necessary? Though it does not specify the reason.

Reply:
It can happen if for unforeseen reasons you get totally removed from staking while you are in the queue (heavy slash).